File: bin/elephants_graveyard/bc-analysis.py
import numpy as np
import pandas as pd
import itertools as it
#import data
# turn this into a loop or function sometime
andata = pd.read_csv("/Users/tristanpwdennis/Projects/bactocap/datasets/anthrax/results/0to400covstats.txt", sep=' ', names=["num_baits", "sample", "depth"]) 
mydata = pd.read_csv("/Users/tristanpwdennis/Projects/bactocap/datasets/mycoplasma/covstatstotal.txt", sep=' ', names=["num_baits", "sample", "depth"])
mymeta = pd.read_csv("/Users/tristanpwdennis/Projects/bactocap-data/mycoplasma/metadata/myco-seq-sample-data.csv")
anmeta = pd.read_csv("/Users/tristanpwdennis/Projects/bactocap-data/anthrax/metadata/anthrax-metadata-1.csv")

#get sampleset identifiers (in this case the number of samples in the set)
#and then get fraction of baits captured to our degree of coverage as a fraction of the total baits for each sampleset
def wrangle(datafile, metadatafile, numbaits): 
    colval = pd.Series([len(metadatafile.index) for i in range(len(datafile.index))])
    datafile.insert(loc = 0, column = 'ninds', value = colval)
    datafile['frac'] = (numbaits - (datafile['num_baits']))/numbaits
    datafile['organism'] = np.where(datafile['ninds']==113, 'anthrax', 'mycoplasma')
    return(datafile)
    
#bind together and add organism names
df = pd.concat([wrangle(mydata, mymeta, 24444), wrangle(andata, anmeta, 148729)])

#intialise empty dataframe
size_df = pd.DataFrame(columns=["frac", "coverage", "ninds", "size"])
#we want to calculate the number of individuals where we covered the genome to a depth of 0X to 400X, and a breadth of 0-1 (as fraction)
#so we create the cartesian product of these ranges, expand them into a triplet in a for loop and iterate over them all (3-way nested loop)
#the count the number of rows matching each combination of conditions, populating a dataframe with the result
for f, c, i in it.product(np.arange(0.1,1.0,0.1), np.arange(0,410,10),[113, 56]):
    mask = (df.frac > f) & (df.depth == c) & (df.ninds == i)
    size = mask.sum()
    size_df.loc[len(size_df)] = [f, c, i, size]
    

#get the fraction of individuals per sample set by dividing the number of individuals matching each condition
size_df['frac_inds'] = size_df['size']/size_df['ninds']


#add organism name for ninds as above (boring and redundant)
size_df['organism'] = np.where(size_df['ninds']==113, 'anthrax', 'mycoplasma')

#plot depth of coverage to which an increasing fraction of baits are totally covered in both
#sample sets

size_df['frac'] = size_df['frac']*100


size_df.to_csv('~/Projects/bactocap/')


pd.read_csv('~/'

File: bin/elephants_graveyard/bc-analysis.R
#####examining coverage over bait sequences for mycoplasma and anthrax

library(dplyr)
library(ggplot2)
library(viridis)
library(data.table)

#get metadata for mycoplasma and anthrax
mycodata <- read.csv("/Users/tristanpwdennis/Projects/bactocap-data/mycoplasma/metadata/myco-seq-sample-data.csv")
anthdata <- read.csv("/Users/tristanpwdennis/Projects/bactocap-data/anthrax/metadata/anthrax-metadata-1.csv")

#remove trailing underscore
#mycoseqdata$sample_id <- sub("_[^_]+$", "", mycoseqdata$sample_id)

#load paths for coverage info
anthraxpath <- c("/Users/tristanpwdennis/Projects/bactocap/datasets/anthrax/results/covstats.txt")
mycopath <- c("/Users/tristanpwdennis/Projects/bactocap/datasets/mycoplasma/covstats.txt")
anthraxhi <- c("/Users/tristanpwdennis/Projects/bactocap/datasets/anthrax/results/200to400covstats.txt")
mycohi <- c("/Users/tristanpwdennis/Projects/bactocap/datasets/mycoplasma/filtered_200to400covstats.txt")

#function to get our files and derive the fraction of bait captures for each coverage level
get_bait_cap_frac <- function(fp, numbaits) {
  cov <- read.csv(fp, sep  = ' ', header = F)
  colnames(cov) <- c('no_baits', 'sampleid', 'coverage')
  cov2 <- mutate(cov, frac = (numbaits - no_baits)/numbaits*100)
  return(cov2)
}

an <- bind_rows(get_bait_cap_frac(anthraxpath, 148729), get_bait_cap_frac(anthraxhi, 148729))
my <- bind_rows(get_bait_cap_frac(mycopath, 24444), get_bait_cap_frac(mycohi, 24444))

ggplot(my, aes(x = coverage, y = frac, color = sampleid)) +
  geom_line(show.legend = FALSE) +
  theme(legend.position = "none") +
  theme_minimal() 




#call function and concatenate results for both anthrax and mycoplasma into single df
concatenated_frac_cov_ind <- rbind(get_noinds_cov(an, anthdata, 'anthrax'), get_noinds_cov(my, mycodata, 'mycoplasma'))

#weird result for 200 as I split the calc over two windows - will rerun and plot again
t<-filter(concatenated_frac_cov_ind, capture_rate != 200)

#plot
ggplot(t, aes(x = capture_rate, y = prop_ind, color = organism)) +
  geom_line()+
  xlab("Depth of coverage over 90% of the captured baits") + ylab("Fraction of Samples") +
  theme_minimal() 



my$organism <- paste0('mycoplasma')
my$ninds <- nrow(mycodata)
an$organism <- paste0('anthrax')
an$ninds <- nrow(anthdata)

comb <- rbind(an, my)

#find the cartesian product of the coverage and genome fraction ranges (every % genome covered at every d.o.c), and organism
covfrac <- expand.grid(seq(10,400, 10), seq (0.01, 1, 0.01), as.numeric(c('56', '117')))

testfunc <- function(se, fracse, numin) {
  x <- nrow(filter(comb, coverage == se & frac > fracse & ninds == numin))/numin
}

results <- as.data.frame(mapply(testfunc, covfrac[,1], covfrac[2], covfrac[,3]))
t <- cbind(results, covfrac)























#percent of samples, per anthrax and mycoplasma, covered to each X

get_noinds_cov <- function(df, metadata, orgname) {
  prop <- data.frame()
  for (i in seq(10,400, 10)) {
    x <- nrow(filter(df, coverage == i & frac > 0.9))/nrow(metadata)
    prop = rbind(prop, x)
  }
  prop <- cbind(prop, as.data.frame(seq(10,400, 10)))
  colnames(prop)<-c('prop_ind', 'capture_rate')
  prop$organism <- paste0(orgname)
  return(prop)
}

